### ISSUE TITLE
Please consider installing `std.cppm` in `lib++`

### ISSUE THREAD

**Issue by boris-kolpackov** [llvm/llvm-project#73089]
I've been trying out the `std` module  in `libc++` with `build2` (per the [Modules in libc++](https://github.com/llvm/llvm-project/blob/main/libcxx/docs/Modules.rst) documentation) and everything seems to work without any issues, at least on simple examples. Very impressive, I must say.

There is just one snag: `std.cppm` is not installed. Quoting the relevant part from the above documentation:

> Once libc++'s implementation is more mature we will reach out to build system vendors, with the goal that building the BMI files is done by the build system.
>
> Currently this requires a local build of libc++ with modules enabled. Since modules are not part of the installation yet, they are used from the build directory.

In `build2` we have the ability to build BMIs of external libraries on the fly (this is needed for consuming modules from installed libraries). So if `std.cppm` were installed somewhere where we could find it, then this would work pretty much out of the box in our case.

I can see two immediate questions about installing `std.cppm`: where to install it and how to make it self-contained (currently, it `#include`s a large number of files from the `std/` subdirectory).

Regarding where to install, I would suggest just installing it next to the headers (e.g., into `/usr/include/c++/v1/`). I can think of two advantage with doing it this way:

1. There is already a way for a build system to find this directory (by parsing the header search paths in the `-v` output) and some build systems already do this (for example, in `build2` we extract the compiler's header and library search paths).

2. The `std.cppm` file will end up in the right distribution package (e.g., `-dev` on Debian, `-devel` on Fedora, etc) without any extra effort.

The only objection I've heard to installing modules next to headers is that "it's not a header" and "it may be included by mistake". While both are true, we are already installing non-header files (e.g., inline/template implementation files) and nobody seems to be including them by mistake. We could also take an extra measure to preclude this by, for example, installing the modules into a subdirectory (e.g., `/usr/include/c++/v1/modules/`).

Regarding making `std.cppm` self-contained, I assume there is no desire to install all the 100+ files from `std/` that it includes (though there is nothing technically bad about doing it, just feels unnecessary). It seems the easiest would be to just append their contents to `std.cppm.in` instead of `#include`ing them. Though I am not sure if this is easy to achieve in CMake.

Thoughts?

P.S. For completeness, let me also mention what I've done as a (hopefully) temporary measure in `build2` to support `import std` right now. I've made a self-contained `std.cppm` (as described above), patched it slightly so that it works for both `libc++` 17 and 18, and bundled it with `build2`. In other words, instead of being installed with `libc++`, it is being installed with `build2` (again, hopefully temporarily and into `build2`'s own directory, naturally).

@mordante @iains @ChuanqiXu9 @dwblaikie @mathstuf

**Comment by ChuanqiXu9** [comment]
FYI, about the position of installing modules, there is a discussion in https://lists.isocpp.org/sg15/2023/07/1971.php

**Comment by boris-kolpackov** [comment]
BTW, to add a data point for the "perhaps the `std` module in `libc++` is mature enough to start installing it" position: I was able to build @ChuanqiXu9 's [`simple_async`'s `CXX20Modules` branch](https://github.com/alibaba/async_simple/tree/CXX20Modules) with `libc++`. All I had to do is fix a few bugs in the project and get rid of one workaround (which is not standard C++). You can see the changes here: https://github.com/boris-kolpackov/async_simple/tree/CXX20Modules-build2

**Comment by mathstuf** [comment]
Note that we don't just need the files, but also the flags to make them work as expected (namely to suppress warnings about reserved module names).

**Comment by ldionne** [comment]
If we start shipping `std.cppm` at a given location, that will become the de-facto standard for build systems to pick up modules files from. I'm not really opposed to doing that but it seems like we'd be stepping on SG15's toes since my understanding is that we haven't come to an official agreement on that yet?

**Comment by mordante** [comment]
Great to hear it works with build2!

Shipping the module files is getting at the top of my todo list and I wanted to investigate which tool vendors to contact. However last week I saw @ruoso created https://github.com/cplusplus/modules-ecosystem-tr/issues/29. So I prefer to participate in that effort instead of starting a similar effort. Build2 is also on that list.

Regarding the large amount of `inc` files. For now I intend to keep them, they are very useful for testing that libc++ exports all named declarations from a header.

**Comment by boris-kolpackov** [comment]
> Note that we don't just need the files, but also the flags to make them work as expected (namely to suppress warnings about reserved module names via -Wno-reserved-module-identifier).

FWIW, this is the only extra option I had to pass to compile `std.cppm`. At least in `build2` this whole logic is quite ad hoc: detect the importation of the `std` modules, notice that it's clang with `libc++`, find `std.cppm`, synthesize the dependency for the BMI, and add the extra option. And I can't see any major generalization in this area. Feels like build systems will just need to handle each compiler/standard library combination. Luckily there aren't that many of them.

> I'm not really opposed to doing that but it seems like we'd be stepping on SG15's toes since my understanding is that we haven't come to an official agreement on that yet?

How can I put this politely? SG15 still hasn't come to an official agreement on the file extension for module but it seems to have agreed on the term used for compiled modules: BMI. Though there is no agreement on what it stands for: some think it's a "binary module interface", some think it's a "built module interface", there is probably another interpretation I am not aware of. So if you wait on SG15, you may end up in a situation where build systems just run out of patience and start bundling your module interfaces.

Also, FWIW, MSVC folks have decided on the location and are guaranteeing it: https://github.com/build2/build2/issues/333#issuecomment-1759932426

> Regarding the large amount of `inc` files. For now I intend to keep them, they are very useful for testing that libc++ exports all named declarations from a header.

Just to clarify, I didn't suggest that you get rid of them entirely, only in the installed version of `std.cppm`. Even that was only a suggestions -- if you decide to install them, I don't think anyone will complain.

**Comment by ChuanqiXu9** [comment]
From https://github.com/cplusplus/modules-ecosystem-tr/issues:

> The first step of this process will be to collect input from a wide range of tooling providers in order to identify requirements in the ecosystem, for that we need to identify who we need to reach out to in order to get the requirements.

So it looks like that issue is not a good place to talk about the installation place for std modules.

If we want to push things forward, maybe we can send another mail to SG15 with a stronger desire to get a conclusion (instead of purely discussion), or if we want, we can make the decision ourselves. (I don't have a strong opinion here)

CC: @mordante @ldionne @boris-kolpackov

**Comment by Bigcheese** [comment]
The specific search paths used to find module source files in general hasn't been something that SG15 has discussed so far. This is largely because that very quickly hits packaging, project layout, and distribution, which are things we care about, but are large issues that would delay getting the minimum needed for modules.

I do think that discussing the limited case of where libc++ and libstdc++ should put their `std` module is something that's useful to discuss in SG15, and I'd be happy to setup a meeting soon to discuss it.

**Comment by boris-kolpackov** [comment]
> If we want to push things forward, maybe we can send another mail to SG15 with a stronger desire to get a conclusion (instead of purely discussion), or if we want, we can make the decision ourselves.

I don't see an issue with making a decision ourselves. MSVC has already decided and if we pick something sensible I don't see a reason why `libstdc++` folks won't adopt it.

I think in the end it doesn't really matter where the modules are installed provided that:

1. It doesn't cause friction for downstream package maintainers in Debian, Fedora, etc. For example, choosing something next to `/usr/include/` such as `/usr/modules/` will definitely cause friction.

2. It's reasonably easy for a build system to determine this location. Paths printed with `-print search-dirs` would be one natural place. Alternatively, something like `-print-file-name=modules` by analogy with `-print-file-name=plugin` in GCC.

The reason I personally like `/usr/include/` is because it's a non-event for downstream maintainers and any serious build system already has to extract header search paths (for example, to suppress duplicates in `pkg-config` options, etc). But I am ok with anything else sensible.

**Comment by boris-kolpackov** [comment]
> I predict that we'll have more arch-specific files than we already have with headers

Do any plausible examples come to mind?

> necessitating some arch-ful bits in `/usr/include`.

There is already a convention for installing arch-specific headers into `/usr/include/<target-triplet>/`, for example into `/usr/include/x86_64-linux-gnu/` on my machine. We've even added an installation location name for this recently: https://github.com/build2/build2/blob/v0.16.0/NEWS#L257-L272

**Comment by mathstuf** [comment]
> Do any plausible examples come to mind?

Anything that does SIMD will have per-arch modules unless the backend is 100% isolated (which seems very unlikely to me) will expose transitive arch-specific modules. Maybe it's resolvable with preprocessor arch detection and things Just Work on Linux? Not sure how this works for macOS universal builds though.

**Comment by ChuanqiXu9** [comment]
> The specific search paths used to find module source files in general hasn't been something that SG15 has discussed so far. This is largely because that very quickly hits packaging, project layout, and distribution, which are things we care about, but are large issues that would delay getting the minimum needed for modules.
> 
> I do think that discussing the limited case of where libc++ and libstdc++ should put their `std` module is something that's useful to discuss in SG15, and I'd be happy to setup a meeting soon to discuss it.

Can we have a meeting in 2 weeks? Since Clang/LLVM18 will be branched in the end of the next month, I really hope we can achieve this before that.

**Comment by Bigcheese** [comment]
> > The specific search paths used to find module source files in general hasn't been something that SG15 has discussed so far. This is largely because that very quickly hits packaging, project layout, and distribution, which are things we care about, but are large issues that would delay getting the minimum needed for modules.
> > I do think that discussing the limited case of where libc++ and libstdc++ should put their `std` module is something that's useful to discuss in SG15, and I'd be happy to setup a meeting soon to discuss it.
> 
> Can we have a meeting in 2 weeks? Since Clang/LLVM18 will be branched in the end of the next month, I really hope we can achieve this before that.

I'll send out a scheduling poll to figure out when the best time to meet is either next week or the week after. It would also be good to have some more detailed info for the other suggestions in the thread (what Boris wrote is a good level of detail for a discussion).

**Comment by mordante** [comment]
> The error towards the end of https://github.com/llvm/llvm-project/blob/main/libcxx/docs/Modules.rst suggests that `std.cppm` itself is not portable across different compiler flags:
> 
> > `error: module file _deps/std-build/CMakeFiles/std.dir/std.pcm cannot be loaded due to a configuration mismatch with the current compilation [-Wmodule-file-config-mismatch]`
> 
> If that's really true (and not just for compiled module representation), it's not a good idea to ship it. Before we can do that, we need something that can be used to produce a compiled module for `-fexceptions`, `-fno-exceptions`, different C++ revisions, and so on.

This is why modules need build system support, they know the user's compilation flags. Then they can create the BMI(s) that match the compiler flags. As @Bigcheese mentioned the source `std.cppm` is portable.

**Comment by boris-kolpackov** [comment]
> Would it be sufficient if the compiler driver had an option to build the std binary module representation according to the other compilation flags?

You would then have to provide a mechanism for detecting when the previously compiled `std` module BMI is out-of-date. Also, presumably, at some point there will be the `std` module in `libstdc++` which we would want to support. Feels like installing the `std` module source and letting the build system take care of the rest is the most sensible approach.

**Comment by boris-kolpackov** [comment]
> Just looking at the file modification time won't cover dependent files even if the source module is materialized on disk.

No, but compiling it like any other translation unit and with header dependency extraction will. It already works for Microsoft's `std.ifc` which includes a bunch of standard library headers.

> That probably needs a more complicated handshake between the projects anyway, and a single source file might not deliver all the required ingredients.

I feel like this discussion is becoming increasingly abstract with mostly hypothetical "what-if" requirements.

**Comment by mordante** [comment]
> > > Would it be sufficient if the compiler driver had an option to build the std binary module representation according to the other compilation flags?
> > 
> > 
> > You would then have to provide a mechanism for detecting when the previously compiled `std` module BMI is out-of-date.
> 
> Dependency tracking options would already provide that kind of information? Just looking at the file modification time won't cover dependent files even if the source module is materialized on disk.
> 
> > Also, presumably, at some point there will be the `std` module in `libstdc++` which we would want to support. Feels like installing the `std` module source and letting the build system take care of the rest is the most sensible approach.
> 
> That probably needs a more complicated handshake between the projects anyway, and a single source file might not deliver all the required ingredients.

There is indeed more involved and it needs coordination between standard libraries, compilers, and build system tools. This is something that's being discussed in SG-15 (the ISO C++ tooling study group).

**Comment by mordante** [comment]
This issue has been discussed in SG-15 and progress was made. So I expect we can have something before branching the next release (the end of January 2024).

**Comment by ChuanqiXu9** [comment]
> This issue has been discussed in SG-15 and progress was made. So I expect we can have something before branching the next release (the end of January 2024).

Could you give a brief summary for the consensus? so that people didn't attend the meeting can be on the same page.

**Comment by boris-kolpackov** [comment]
Let me copy what I believe is the relevant reply from the SG15 mailing list:

> > In my mind, the motivation of the meeting should be the location of the std modules. Should it be `<PREFIX>/modules/c++/v1`? Should it be `/usr/include/c++/$version`?  Or should it be `$PREFIX/share/c++/modules`?
>
> The consensus is that it should be intermediated by the metadata file, which makes the specific location less important, and open for package managers to customize.
>
> There was also a brief conversation where the feeling was that we shouldn't reuse the include directory for this, and instead this should be inside a private location for the standard library.

So my summary is SG15 added another level of indirection (the metadata file), but where to install it or the actual module interface units as well as how to convey this information to build systems is left as an exercise to the implementers.

**Comment by iains** [comment]
> Let me copy what I believe is the relevant reply from the SG15 mailing list:
> 
> > > In my mind, the motivation of the meeting should be the location of the std modules. Should it be `<PREFIX>/modules/c++/v1`? Should it be `/usr/include/c++/$version`?  Or should it be `$PREFIX/share/c++/modules`?
> > 
> > 
> > The consensus is that it should be intermediated by the metadata file, which makes the specific location less important, and open for package managers to customize.
> > There was also a brief conversation where the feeling was that we shouldn't reuse the include directory for this, and instead this should be inside a private location for the standard library.
> 
> So my summary is SG15 added another level of indirection (the metadata file), but where to install it or the actual module interface units as well as how to convey this information to build systems is left as an exercise to the implementers.

Another way of looking at this is that SG15 is trying to determine a mechanism (and actually to specify the content of the metadata file) but giving maximum freedom to the toolchain and OS implementations in how they wish to lay things out.

(my take on) The underlying theme is
 - that the toolchain knows how to find libraries (whether that is done by the driver or some other sub-tool)
 - the relevant metadata files need to be tied to a library ABI (in the case of toolchains supporting  multilibs, for example). So that the toolchain is already making that distinction (in order to link the correct ABI library) and should be able to make the same distinction for the metadata.

perhaps a reasonable initial implementation would be to place the metadata alongside each library ABI slice (and then the same driver/sub-tool process can be used to locate it based on ABI-changing flags [e.g. hard/soft float or 32 / 64B ISA]).

Please make input on the SG15 reflector in helping figure out the initial metadata content, and in comments/improvements to the scheme in general (it must still be considered Work In Progress).

**Comment by boris-kolpackov** [comment]
> Another way of looking at this is that SG15 is trying to determine a mechanism (and actually to specify the content of the metadata file) but giving maximum freedom to the toolchain and OS implementations in how they wish to lay things out.

From the build system's POV (which is the primary consumer of this mechanism), it doesn't make a difference whether the compiler gives it the location of the standard modules directly or it give it the location of the metadata file from which the build system gets those locations. If anything, it's an extra hoop to jump through. I do understand that this metadata file may also provide extra information (compile options, etc) that is necessary to build the modules, but it is not the case currently. More precisely, nothing extra is needed for compiling MS STL modules and for `libc++` only `-Wno-reserved-module-identifier` is needed, which, at least in case of `build2`, is a lot easier to add ad hoc than to parse a JSON file. So from my POV, SG15 is focusing on theoretical problems that we currently don't have (but may have in the future, admittedly) while ignoring practical problems that we do have (like helping implementers decide where to install module interfaces and how to communicate this to the build systems).

However,

> This issue has been discussed in SG-15 and progress was made. So I expect we can have something before branching the next release (the end of January 2024).

this is all that really matters. If the SG15 discussion helped and we can start using `import std;` in Clang 18, I am happy to hear that. I can parse an extra JSON file or two.

**Comment by Bigcheese** [comment]
> So my summary is SG15 added another level of indirection (the metadata file), but where to install it or the actual module interface units as well as how to convey this information to build systems is left as an exercise to the implementers.

The metadata file is needed because the flags needed to build the module interface for the std module cannot be determined by looking at the input command line unless you want to hard code that information in every build system. For example you need the extra flag to say that you're allowed to name a module `std`, and for the MS stdlib you need to say don't treat anything as header units, even if the importer is.

Additionally for ABI breaking differences, the options you need are directly tied to the library binary being used, not to the source code the interface was built from. 

Clang and GCC currently plan to use the already existing `-print-file-name` to provide the path to this file to the build system.

**Comment by ChuanqiXu9** [comment]
While I respect the work of SG15, will it be a  blocking issue for shipping std modules in libc++?

I mean, for example, can we use the proposal from Daniel as the first step?

```
{
  "version": 1,
  "revision": 1,
  "modules": [
    {
      "logical-name": "std",
      "source-path": "modules/std.cppm",
      "is-standard-library": true
    },
    {
      "logical-name": "std.compat",
      "source-path": "modules/std.compat.cppm"
      "is-std-library": true
    },
  ]
}
```

It doesn't solve the issue you raised but it allows us to ship module units initially so that people can start experimenting and get an initial feeling. Also I think it won't add extra works in this direction (add flags to the metadata file).

**Comment by mordante** [comment]
> While I respect the work of SG15, will it be a blocking issue for shipping std modules in libc++?

Quite the opposite, the discussion in SG15 gave a good direction for a way forward.

> I mean, for example, can we use the proposal from Daniel as the first step?

Yes. I plan to look at that soon so build systems can use it.

> > The metadata file is needed because the flags needed to build the module interface for the std module cannot be determined by looking at the input command line unless you want to hard code that information in every build system. For example you need the extra flag to say that you're allowed to name a module `std`, and for the MS stdlib you need to say don't treat anything as header units, even if the importer is.
> 
> I still don't see why you couldn't use a simple flag that tells the compiler to write out the binary object representation matching the rest of the command line flags. The build system already needs compiler-specific knowledge how to set the output path, how to request compilation instead of linking, and so on, so knowledge of this new flag seems a rather minor addition. In this approach, is up to the compiler to get itself into a state were it is able to create a module named `std`, so a whole bunch of potential issues simply do not apply.

It would require the compiler to know things it currently doesn't know. For example when the vendor builds libc++ without support for exceptions the compiler needs to issue a `-fno-exception` flag when building the modules. This is not known when the compiler is being built. Having a separate file with the meta data reduces the coupling between the compiler and the library.

**Comment by boris-kolpackov** [comment]
> I mean, for example, can we use the proposal from Daniel as the first step?

FWIW, this metadata file would be sufficient for my needs. I have no problem hardcoding `-Wno-reserved-module-identifier`, which is the least of my problems in this area.

> For example when the vendor builds libc++ without support for exceptions the compiler needs to issue a `-fno-exception` flag when building the modules.

This is actually an interesting question. Let's say libc++ was compiled with `-fno-exceptions` while an application that tries to use it is not. Presumable this should result in an error at some point or maybe not? If there should be an error, I can see two way it can play out:

1. The build system compiles `std` with `-fno-exceptions` (from the metadata file). Then when this BMI is used to compile an application's translation unit, there is presumably an error saying the BMI was built with incompatible options.

2. The build system compiles `std` with flags of the application (and thus without `-fno-exceptions`). Do we expect this to succeed? If so, will there be an error at some later stage (seems like it can only happen during linking)?

From the user's perspective, it seems the ideal outcome would be to fail when building the `std` module BMI. Not sure if it's practically doable though. Maybe this was all covered in the SG15 discussion, in which case apologies for the noise.

**Comment by iains** [comment]
> > I mean, for example, can we use the proposal from Daniel as the first step?
and that is related to / takes into account the existing practice for MSVC.

> > For example when the vendor builds libc++ without support for exceptions the compiler needs to issue a `-fno-exception` flag when building the modules.
> 
> This is actually an interesting question. Let's say libc++ was compiled with `-fno-exceptions` while an application that tries to use it is not. Presumable this should result in an error at some point or maybe not? I

I think that depends on whether there are two instances of the library binary, one with and one without exceptions.  If that is the case, then the user's flags would be used to select which one (I'd assume) and therefore there should be two metadata files one for each incompatible library build.

If there is only one build (e.g. without exceptions) then, presumably the `-fno-exceptions` flags should be added from the metadata (i.e. as a 'required build flag').  An alternate is that the library headers emit a diagnostic in the case that the user's flags conflict [since the presence of that flag is usually reflected in some cpp define] (i.e. we assume that the person building without exceptions did that in a deliberate way that can be detected during builds).

**Comment by boris-kolpackov** [comment]
> I think that depends on whether there are two instances of the library binary, one with and one without exceptions. If that is the case, then the user's flags would be used to select which one (I'd assume) and therefore there should be two metadata files one for each incompatible library build.

Right, though I was specifically drilling down on the case where the standard library flags are incompatible with the user flags. But this suggests a third option in my list above:

3. The build system invokes the compiler with `-print-file-name=std-modules` but without `-fno-exception`, the compiler see that there is no suitable version of the standard library, and returns nothing. The build system then reports to the user that no suitable version of the standard library is available for their build configuration. I like that, if we can swing it.

**Comment by mordante** [comment]
I've created an RFC PR (#75741) for the libc++ changes.

**Comment by boris-kolpackov** [comment]
@mordante Should this issues be reopened seeing that the commit in question has been reverted?

Also, I see that Clang 18 rc1 has been published which makes me wonder if the ship has sailed on having this functionality in 18. Or are you still planning/hoping to get it in before the final release?

**Comment by boris-kolpackov** [comment]
> I don't think we should reopen this since the revered patch is in the clang part. And the job of the libc++ part, to provide option to install it, is completed.

This feels like splitting hairs to me since there is little use of the installed `libc++` modules if nobody can find them. And to me this issue was nicely tracking the overall functionality (compared to a number of RFCs and PRs that were generated to address this) that would be needed by a build system. But Ok.

**Comment by boris-kolpackov** [comment]
I am trying to understand (and document) what ended up making it into Clang 18 specifically looking at Clang 18.1.5 in Debian (the latest release at the time of writing).

The `-print-library-module-manifest-path` option, it appears, did not make it:

```
clang++-18 -std=c++23 -stdlib=libc++ -print-library-module-manifest-path
clang++-18: error: unknown argument: '-print-library-module-manifest-path'
clang++-18: error: no input files
```

This seems to match https://github.com/llvm/llvm-project/pull/85637, which is an attempt to backport this to 18 that was abandoned.

However, the JSON file and the `std`/`std.compat` source files appear to be present:

```
$ cat /usr/lib/llvm-18/lib/libc++.modules.json 
{
  "version": 1,
  "revision": 1,
  "modules": [
    {
      "logical-name": "std",
      "source-path": "../share/libc++/v1/std.cppm",
      "is-std-library": true,
      "local-arguments": {
        "system-include-directories": [
          "../share/libc++/v1"
        ]
      }
    },
    {
      "logical-name": "std.compat",
      "source-path": "../share/libc++/v1/std.compat.cppm",
      "is-std-library": true,
      "local-arguments": {
        "system-include-directories": [
          "../share/libc++/v1"
        ]
      }
    }
  ]
}
```

```
ls -1 /usr/lib/llvm-18/lib/../share/libc++/v1/
std/
std.compat/
std.compat.cppm
std.cppm
```

Is this understanding accurate or am I missing something? If it is accurate, is it also accurate to assume that the full support for this will only be available in Clang 19?

**Comment by ChuanqiXu9** [comment]
> I am trying to understand (and document) what ended up making it into Clang 18 specifically looking at Clang 18.1.5 in Debian (the latest release at the time of writing).
> 
> The `-print-library-module-manifest-path` option, it appears, did not make it:
> 
> ```
> clang++-18 -std=c++23 -stdlib=libc++ -print-library-module-manifest-path
> clang++-18: error: unknown argument: '-print-library-module-manifest-path'
> clang++-18: error: no input files
> ```
> 
> This seems to match #85637, which is an attempt to backport this to 18 that was abandoned.
> 
> However, the JSON file and the `std`/`std.compat` source files appear to be present:
> 
> ```
> $ cat /usr/lib/llvm-18/lib/libc++.modules.json 
> {
>   "version": 1,
>   "revision": 1,
>   "modules": [
>     {
>       "logical-name": "std",
>       "source-path": "../share/libc++/v1/std.cppm",
>       "is-std-library": true,
>       "local-arguments": {
>         "system-include-directories": [
>           "../share/libc++/v1"
>         ]
>       }
>     },
>     {
>       "logical-name": "std.compat",
>       "source-path": "../share/libc++/v1/std.compat.cppm",
>       "is-std-library": true,
>       "local-arguments": {
>         "system-include-directories": [
>           "../share/libc++/v1"
>         ]
>       }
>     }
>   ]
> }
> ```
> 
> ```
> ls -1 /usr/lib/llvm-18/lib/../share/libc++/v1/
> std/
> std.compat/
> std.compat.cppm
> std.cppm
> ```
> 
> Is this understanding accurate or am I missing something? If it is accurate, is it also accurate to assume that the full support for this will only be available in Clang 19?

Yes, I think this is accurate. BTW, maybe it is helpful to mention that, whether or not the json file gets installed depends on vendors choices.

CC @mordante and @mathstuf for double check.

**Comment by boris-kolpackov** [comment]
Another data point: I've installed Clang 19 snapshot from Debian experimental:

```
Debian clang version 19.0.0 (++20240421021844+e095d978ba47-1~exp1)
```

In this version the `-print-library-module-manifest-path` option is recognized but it doesn't print anything sensible:

```
$ clang++-19 -std=c++23 -stdlib=libc++ -print-library-module-manifest-path
<NOT PRESENT>
```

Even though the JSON file and the `std`/`std.compat` source files are installed, the same as for Clang 18.

**Comment by ChuanqiXu9** [comment]
> Another data point: I've installed Clang 19 snapshot from Debian experimental:
> 
> ```
> Debian clang version 19.0.0 (++20240421021844+e095d978ba47-1~exp1)
> ```
> 
> In this version the `-print-library-module-manifest-path` option is recognized but it doesn't print anything sensible:
> 
> ```
> $ clang++-19 -std=c++23 -stdlib=libc++ -print-library-module-manifest-path
> <NOT PRESENT>
> ```
> 
> Even though the JSON file and the `std`/`std.compat` source files are installed, the same as for Clang 18.

Oh, this should be a bug IIRC. @mordante. Maybe we didn't receive bug reports just because no one uses it :  (

Can you compile a hello world example with `-std=libc++` in this configuration?

**Comment by mathstuf** [comment]
The `-print-file-name=libc++.modules.json` flag works for CMake to find the file.

**Comment by boris-kolpackov** [comment]
> Can you compile a hello world example with `-std=libc++` in this configuration?

Yes. I can even do one better and compile it with `import std;` if I connect things manually.

> The `-print-file-name=libc++.modules.json` flag works for CMake to find the file.

Hm, yes, this works for me also with Clang 19 (but not 18).

Now I am wondering what is the difference/purpose of `-print-library-module-manifest-path`? Can someone give an authoritative answer as to what mechanism a build system vendor should use?

**Comment by mathstuf** [comment]
I tested it with a custom-built 18.1.2, so why 18 doesn't work in your case is an interesting question…

**Comment by ChuanqiXu9** [comment]
> Now I am wondering what is the difference/purpose of -print-library-module-manifest-path? Can someone give an authoritative answer as to what mechanism a build system vendor should use?

Personally I think the intention should be `-print-library-module-manifest-path`. But it might be better to ask @mordante

**Comment by mordante** [comment]
The manifest is indeed not installed by default in Clang 18. It will be installed by default in Clang 19. This still requires Linux distributions to add these files to their packages or provide additional packages.

The intention was indeed to use `-print-library-module-manifest-path`, but this indeed seems to be broken :-( It's odd, it used to work on my system. The intention is that the file could have a different name depending on compilation options.

**Comment by boris-kolpackov** [comment]
> I tested it with a custom-built 18.1.2, so why 18 doesn't work in your case is an interesting question…

I also checked Clang in 18.1.1 that comes in Fedora 40:

1. `-print-library-module-manifest-path` is not recognized.

2. `-print-file-name=libc++.modules.json` doesn't print a path.

3. Neither `libc++.modules.json` nor `std`/`std.compat` source files appear to be present.

**Comment by mordante** [comment]
Based on `Neither libc++.modules.json nor std/std.compat source files appear to be present.` it seems these files are not installed on Fedora.

**Comment by boris-kolpackov** [comment]
Right, we seems to have a full spectrum of ways in which this is not working. FWIW, we are shipping the hack in `build2` `0.17.0` since it is currently the only way to make `import std;` work with Clang 18. I hope this can be fixed in time for Clang 19. But I have my doubts about making sure things are installed consistently across major distributions.

**Comment by boris-kolpackov** [comment]
@ldionne, @mordante: Is there a new milestone for making this work? Perhaps the Clang 19 release?

**Comment by ldionne** [comment]
I haven't followed the whole thread here, but our intention was for this to work in LLVM 19 out of the box. I believe we don't expect anything to work well with the 18 release. Libc++ should be fulfilling its part of the contract here, we do install the module files by default. That still requires the distributions to actually install those files as well.

If Clang misbehaves when the module files are installed properly, then that would be a Clang bug.

**Comment by ChuanqiXu9** [comment]
Let me try to give a summary since it is too long and I am not sure if I understand correctly.

1. If `libc++.modules.json` or std/std.compat source doesn't appear, it should be the that vendors doesn't provide that.

2. If `-print-file-name=libc++.modules.json` works? If yes, then it should be the problem of our patch to implement `-print-library-module-manifest-path` and if not, there might be other problems.

Also in the case `2` is true, i suggest to file a new issue for this since it looks like a different topic than this thread.

**Comment by boris-kolpackov** [comment]
@ChuanqiXu9 I personally still don't understand whether it is `-print-library-module-manifest-path` or `-print-file-name=libc++.modules.json` that we are aiming for. @mordante suggests in an earlier comment that it's the former:

> The intention was indeed to use `-print-library-module-manifest-path`, but this indeed seems to be broken

You seem to suggest that it's the latter (actually, you use one option in the first sentence, and the other in the second, so I am not really sure what you are suggesting, if anything).

I upgraded to the latest Clang/libc++ 19 pre-release from Debian experimental:

```
Debian clang version 19.0.0 (++20240529093513+2cfea14a57ad-1~exp1)
```

The modules are installed:

```
$ ls -1 /usr/lib/llvm-19/lib/../share/libc++/v1/
std
std.compat
std.compat.cppm
std.cppm
```

The `-print-library-module-manifest-path` option is recognized but does not work:

```
$ clang++-19 -std=c++23 -stdlib=libc++ -print-library-module-manifest-path
<NOT PRESENT>
```

The `-print-file-name=libc++.modules.json` option is recognized and appears to work:

```
$ clang++-19 -std=c++23 -stdlib=libc++ -print-file-name=libc++.modules.json
/usr/lib/llvm-19/bin/../lib/libc++.modules.json
```

So let me repeat my earlier request: 

_Can someone give an authoritative answer as to what mechanism a build system vendor should use?_ 

We are almost 60 comments down in this issue, and we still can't seem to figure this out!

**Comment by ChuanqiXu9** [comment]
> Can someone give an authoritative answer as to what mechanism a build system vendor should use?

We said, it should be `-print-library-module-manifest-path` in the end of the day. But it is broken now. And in case `-print-file-name=` works, it looks fine to use that first. When we fix the problem, it will be suggested to use  `-print-library-module-manifest-path` again.

**Comment by boris-kolpackov** [comment]
@ChuanqiXu9 Ok, thanks for the clarification. Could you also explain what would be the incentive for me as a build system vendor to later switch from ` -print-file-name=libc++.modules.json` to `-print-library-module-manifest-path`? It would have to be something substantial to risk destabilizing things. Maybe it makes more sense to just kill `-print-library-module-manifest-path` and call it a day?

**Comment by ChuanqiXu9** [comment]
> @ChuanqiXu9 Ok, thanks for the clarification. Could you also explain what would be the incentive for me as a build system vendor to later switch from ` -print-file-name=libc++.modules.json` to `-print-library-module-manifest-path`? It would have to be something substantial to risk destabilizing things. Maybe it makes more sense to just kill `-print-library-module-manifest-path` and call it a day?

To be honest, for my memory, `-print-library-module-manifest-path` is proposed  during a SG15 meeting. So it stands for the consensus of SG15. So in the **ideal** world, if build system vendors follow this, they should be able to avoid some if-else to calculate the path to libc++. And if build systems don't do this, they just need to add some `if-else` as far as I know.

**Comment by boris-kolpackov** [comment]
Ok, some new exciting developments in this saga:

```
$ clang++-20 --version
Debian clang version 20.1.6 (1~exp1)

$ clang++-20 -print-file-name=libc++.modules.json
/usr/lib/llvm-20/bin/../lib/libc++.modules.json

$ clang++-20 -print-file-name=libstdc++.modules.json
/usr/lib/gcc/x86_64-linux-gnu/15/libstdc++.modules.json

$ clang++-20 -stdlib=libstdc++ -print-library-module-manifest-path
/usr/lib/gcc/x86_64-linux-gnu/15/libstdc++.modules.json

$ clang++-20 -stdlib=libc++ -print-library-module-manifest-path
<NOT PRESENT>
```

So looks like `-print-library-module-manifest-path` works but only for `libstdc++`, not `libc++`.

**Comment by kamrann** [comment]
To add some more datapoints:

On Fedora-41 via WSL, distro installed clang:
```
$ clang++ --version
clang version 19.1.5 (Fedora 19.1.5-1.fc41)

$ clang++ -print-file-name=libc++.modules.json
/usr/lib64/libc++.modules.json

$ clang++ -stdlib=libc++ -print-library-module-manifest-path
/usr/lib64/libc++.modules.json

$ clang++ -print-file-name=libstdc++.modules.json
libstdc++.modules.json

$ clang++ -stdlib=libstdc++ -print-library-module-manifest-path
<NOT PRESENT>
```

On Fedora-41 via WSL, source built non-installed clang trunk:
```
$ clang++ --version
clang version 21.0.0git (https://github.com/llvm/llvm-project.git 0d68bad78a68874c31cdb337f97a0c4336b1125b)

$ clang++ -print-file-name=libc++.modules.json
/home/.../llvm/out/lib/x86_64-unknown-linux-gnu/libc++.modules.json

$ clang++ -stdlib=libc++ -print-library-module-manifest-path
/home/.../llvm/out/lib/x86_64-unknown-linux-gnu/libc++.modules.json

$ clang++ -print-file-name=libstdc++.modules.json
libstdc++.modules.json

$ clang++ -stdlib=libstdc++ -print-library-module-manifest-path
<NOT PRESENT>
```

So in both of these cases, working as expected for libc++ but not for libstdc++.

On Windows itself. Not sure if this is expected to work, I should add that I haven't tried to get libc++ working on Windows so it's likely excluded from my LLVM builds. And no idea if this command should be capable of detecting MSSTL:
```
$ clang++ --version
clang version 20.1.2 (https://github.com/llvm/llvm-project.git 41c36d94080488cc938b1c1697c7e8353405cd75)

$ clang++ -print-file-name=libc++.modules.json
libc++.modules.json

$ clang++ -stdlib=libc++ -print-library-module-manifest-path
<NOT PRESENT>

$ clang++ -print-library-module-manifest-path
<NOT PRESENT>
```

Note that where `-print-file-name` fails to give a full path, it's just printing back out whatever was given as the argument. So `-print-library-module-manifest-path` seems more reliable from my tests, but that doesn't match with @boris-kolpackov 's result above where it fails in a case that `-print-file-name` works.

**Comment by ChuanqiXu9** [comment]
BTW, like I said in https://github.com/llvm/llvm-project/issues/140825,  for build system guys, I do suggesting you to try to contribute in driver part. Since the code in driver part is independent form the rest of the repo. The code structure is more or less "trivial". And the only domain knowledge it requires are how compiler flags should be combined to work, which is the area for build system.

**Comment by boris-kolpackov** [comment]
FWIW, GCC 15 does not recognize `-print-library-module-manifest-path` as a valid option, only `-print-file-name`.

**Comment by boris-kolpackov** [comment]
I see Clang is installed from https://apt.llvm.org/. So the first thing to check is whether the `libc++-dev` package contains the module. If not, then I would file a bug with the package maintainer and they could forward it to upstream (if it turns out to be upstream's problem). At least this is Debian's standard bug escalation process.
